### PR TITLE
Add unit test naming conventions

### PR DESCRIPTION
**Pull request by coroiu**
## Objective

Migrate the WIP document about naming conventions from confluence

Confluence page can be found here: https://bitwarden.atlassian.net/wiki/spaces/DEV/pages/312737793/WIP+Unit+test+conventions

### CODE REVIEW

File: docs/contributing/testing/naming-conventions.mdx
Comment (withinfocus) [inline docs/contributing/testing/naming-conventions.mdx:8]: ❓ Is placing this common practice? I support it being implied and therefore available for future improvement at any time, so this wouldn't need to be here. It'll be as official as this PR review will make i! 😉

If it's because of your "WIP" content below and you're not gonna merge it as-is, I get it. Could be worthwhile to leave the PR as a draft too.

File: docs/contributing/testing/naming-conventions.mdx
Comment (withinfocus) [inline docs/contributing/testing/naming-conventions.mdx:1]: ⛏️ Capital C for level one headings.

File: docs/contributing/testing/naming-conventions.mdx
Comment (withinfocus) [inline docs/contributing/testing/naming-conventions.mdx:17]: 💭 Do these references need to be here, or just in the justification in the PR that adds this definition? I am in the latter camp as you can trace why and how a documentation change was made, but we wouldn't have to justify in the documentation itself like a paper with references.

File: docs/contributing/testing/naming-conventions.mdx
Comment (withinfocus) [inline docs/contributing/testing/naming-conventions.mdx:65]: ⛏️ `jest`.

File: docs/contributing/testing/naming-conventions.mdx
Comment (withinfocus) [inline docs/contributing/testing/naming-conventions.mdx:95]: ⛏️ Colon at the end.

File: docs/contributing/testing/naming-conventions.mdx
Comment (withinfocus) [inline docs/contributing/testing/naming-conventions.mdx:200]: ℹ️ See earlier comment.

File: docs/contributing/testing/naming-conventions.mdx
Comment (withinfocus) [inline docs/contributing/testing/naming-conventions.mdx:10]: ❓ After I read this I got a pretty clear intent for `jest`-based tests but what about our others, like with xUnit? Should this be segmented on framework a bit, and how much can we apply generally but then share framework-specific examples?

File: docs/contributing/testing/naming-conventions.mdx
Comment (withinfocus) [reply docs/contributing/testing/naming-conventions.mdx:25]: Per my earlier comment about framework-specific examples, I often default to my C# experience and the Microsoft examples of roughly "given-when-then". Their example:

`Add_MultipleNumbers_ReturnsSumOfNumbers`

isn't precisely that but connects with @coroiu's (1) above.

Are Bitwarden engineers finding a lot of prerequisites defined for tests that wouldn't fit into "state"?

File: docs/contributing/testing/naming-conventions.mdx
Comment (coroiu) [reply docs/contributing/testing/naming-conventions.mdx:8]: I initially wanted this PR to be a pure migration of the same document in confluence, but you have a good point and I think this should be removed!

File: docs/contributing/testing/naming-conventions.mdx
Comment (coroiu) [reply docs/contributing/testing/naming-conventions.mdx:25]: To be honest I would've suggested `MethodName_StateUnderTest_ExpectedBehavior` if it wasn't so different from what we already have in this repo, it's my favorite convention from my C#-heavy days. 

I actually prefer `<unit> [given -] when - then` because I think it's so similar to `MethodName_StateUnderTest_ExpectedBehavior`. It's basically just the last two flippe `<unit = MethodName> should <ExpectedBehavior> when <StateUnderTest>`

File: docs/contributing/testing/naming-conventions.mdx
Comment (coroiu) [reply docs/contributing/testing/naming-conventions.mdx:10]: I had big plans on expanding this document to other languages, working together with @fedemkr to encompass phone apps, and @eliykat and @differsthecat to expand into other areas such as "How to test frontend components" etc. But I always got the feeling that this was a bit outside my responsibilities and other work just got prioritized in the end. @differsthecat talk about testing reminded me of this earlier this week and now here we are :D

File: docs/contributing/testing/naming-conventions.mdx
Comment (coroiu) [reply docs/contributing/testing/naming-conventions.mdx]: Fixed

File: docs/contributing/testing/naming-conventions.mdx
Comment (coroiu) [reply docs/contributing/testing/naming-conventions.mdx]: Removed!

File: docs/contributing/testing/naming-conventions.mdx
Comment (coroiu) [reply docs/contributing/testing/naming-conventions.mdx]: Fixed!

File: docs/contributing/testing/naming-conventions.mdx
Comment (coroiu) [reply docs/contributing/testing/naming-conventions.mdx]: Fixed!

File: docs/contributing/testing/naming-conventions.mdx
Comment (coroiu) [reply docs/contributing/testing/naming-conventions.mdx]: Removed!

**Review by withinfocus — APPROVED** [review]
Awesome stuff -- not sure if you want to wait on @eliykat or not and we keep iterating. I could benefit from this new testing section now since I am gonna document load tests.

File: docs/contributing/testing/naming-conventions.mdx
Comment (coroiu) [reply docs/contributing/testing/naming-conventions.mdx:77]: How did that happen lol, I'll fix!

File: docs/contributing/testing/index.md
Comment (coroiu) [reply docs/contributing/testing/index.md:1]: Removed the file completely

File: docs/contributing/testing/naming-conventions.mdx
Comment (coroiu) [reply docs/contributing/testing/naming-conventions.mdx]: Fixed, removed `should` as per previous discussion with @eliykat in confluence

File: docs/contributing/testing/naming-conventions.mdx
Comment (coroiu) [reply docs/contributing/testing/naming-conventions.mdx:169]: Fixed, removed `should` as per previous discussion with @eliykat in confluence

File: docs/contributing/testing/index.md
Comment (coroiu) [reply docs/contributing/testing/index.md:1]: I've also restructured a bit, splitting the document in two, creating a new "test structure" document which is still IWP

File: docs/contributing/testing/naming-conventions.mdx
Comment (coroiu) [reply docs/contributing/testing/naming-conventions.mdx]: NOTE: As per previous discussion with @eliykat `should` is no longer part of the name.

**Comment by coroiu** [comment]
@withinfocus In @eliykat last comment on confluence regarding the actual name convention he said

> I don’t think I want to get too hung up on it.

So as soon as I get approval from @Hinton I'll merge the PR

**Review by Hinton — APPROVED** [review]

**Pull request merged**
